### PR TITLE
org.libreoffice.LibreOffice.BundledExtension.Voikko

### PR DESCRIPTION
**Pull request by stbergmann**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by nedrichards** [comment]
bot, build org.libreoffice.LibreOffice.BundledExtension.Voikko

File: org.libreoffice.LibreOffice.BundledExtension.Voikko.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
+{
+    "id": "org.libreoffice.LibreOffice.BundledExtension.Voikko",
+    "runtime": "org.libreoffice.LibreOffice",
+    "runtime-version": "stable",
+    "sdk": "org.freedesktop.Sdk//20.08",
+    "build-extension": true,
+    "separate-locales": false,
+    "appstream-compose": false,
```
Comment (hfiguiere) [inline org.libreoffice.LibreOffice.BundledExtension.Voikko.json:8]:
```suggestion
    "appstream-compose": false,
    "build-options": {
        "prefix": "/app/libreoffice/share/extensions/Voikko",
        "prepend-path": "/app/libreoffice/share/extensions/Voikko/bin"
    },
```

File: org.libreoffice.LibreOffice.BundledExtension.Voikko.json [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            "name": "libvoikko",
+            "buildsystem": "simple",
+            "build-commands": [ "./configure --prefix=/app/libreoffice/share/extensions/Voikko/build-deps/libvoikko --disable-hfst --with-dictionary-path=/app/libreoffice/share/extensions/Voikko/voikko",
+                                "make",
+                                "make install",
+                                "mkdir -p /app/libreoffice/share/extensions/Voikko/voikko/Linux",
+                                "cp -L /app/libreoffice/share/extensions/Voikko/build-deps/libvoikko/lib/libvoikko.so.1 /app/libreoffice/share/extensions/Voikko/voikko/Linux/",
+                                "mkdir -p /app/libreoffice/share/extensions/Voikko/pythonpath",
+                                "cp python/libvoikko.py /app/libreoffice/share/extensions/Voikko/pythonpath/" ],
```
Comment (hfiguiere) [inline org.libreoffice.LibreOffice.BundledExtension.Voikko.json:19]:
```suggestion
            "buildsystem": "autoconf",
            "config-opts": [
                "--disable-hfst",
                "--with-dictionary-path=${FLATPAK_DEST}/voikko"
            ],
            "post-install": [ 
                "install -d ${FLATPAK_DEST}/pythonpath",
                "cp python/libvoikko.py ${FLATPAK_DEST}/pythonpath/" 
            ],
            "cleanup": [
                "/bin"
            ],
```

File: org.libreoffice.LibreOffice.BundledExtension.Voikko.json [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                {
+                    "type": "patch",
+                    "path": "foma-variables.patch"
+                }
+            ]
+        },
+        {
+            "name": "voikko-fi",
+            "buildsystem": "simple",
+            "build-commands": [ "PATH=$PATH:/app/libreoffice/share/extensions/Voikko/build-deps/libvoikko/bin:/app/libreoffice/share/extensions/Voikko/build-deps/foma/bin make vvfst", "make vvfst-install DESTDIR=/app/libreoffice/share/extensions/Voikko/voikko" ],
```
Comment (hfiguiere) [inline org.libreoffice.LibreOffice.BundledExtension.Voikko.json:51]:
```suggestion
            "build-commands": [ 
                "make vvfst", 
                "make vvfst-install" 
            ],
```

File: org.libreoffice.LibreOffice.BundledExtension.Voikko.json [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                    "sha256": "320b2d4e428f6beba9d0ab0d775f8fbe150284fbbafaf3e5afaf02524cee28cc"
+                }
+            ]
+        },
+        {
+            "name": "libreoffice-voikko",
+            "buildsystem": "simple",
+            "build-commands": [ "make oxt",
+                                "mkdir -p /app/libreoffice/share/extensions/Voikko",
+                                "unzip -d /app/libreoffice/share/extensions/Voikko build/voikko.oxt" ],
```
Comment (hfiguiere) [inline org.libreoffice.LibreOffice.BundledExtension.Voikko.json:65]:
```suggestion
                                "unzip -d ${FLATPAK_DEST} build/voikko.oxt" ],
```

File: org.libreoffice.LibreOffice.BundledExtension.Voikko.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                    "sha256": "8322b58d83eb6e1398d6914885d88a2ee08c8dd2fc2b72d75fba8fe83eefbe38"
+                },
+                {
+                    "type": "patch",
+                    "path": "libreoffice-voikko-path.patch"
+                }
+            ]
+        }
+    ],
+    "cleanup-commands": [ "rm -r /app/libreoffice/share/extensions/Voikko/build-deps" ]
```
Comment (hfiguiere) [inline org.libreoffice.LibreOffice.BundledExtension.Voikko.json:79]:
```suggestion
```

File: flathub.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "skip-arches": ["arm"]
```
[the remark this answers is not in the document]
Comment (stbergmann) [reply flathub.json]: Where do you see a flathub.json file in <https://github.com/flathub/flathub/commit/e2aaf7193988e19304d5e036ef5db57f924c99e0> "Provide Voikko as a bundled extension for LibreOffice"?  (Note that I had erroneously pushed that first as <https://github.com/flathub/flathub/commit/d1c7539fa9e340a59d5d435310116a90b4ba9878> "Provide Voikko as a bundled extension for LibreOffice" based on stale content from an old stbergmann/flathub branch, and then force-pushed the correction.)

**Comment by stbergmann** [comment]
@hfiguiere, thanks for the suggestions.  I liked the original way of clearly separating the build-time dependencies under ${FLATPAK_DEST}/build-deps/ instead of mangling their artefacts into ${FLATPAK_DEST} itself (and whose final content has a radically different directory structure from the traditional Unix bin, include, lib, share structure).  But sure, it simplifies some of the build script lines, and if you think that's better, so be it.  Incorporated your suggestions, with necessary fixes, in <https://github.com/flathub/flathub/commit/7784e57d8edcb2b81c462d234b65b52e4b19b48d> "Provide Voikko as a bundled extension for LibreOffice" (and added the "The upstream pythonpath/VoikkoHandlePool.py..." paragraph to the commit message, which I had originally forgotten).

**Comment by hfiguiere** [comment]
bot, build org.libreoffice.LibreOffice.BundledExtension.Voikko

**Comment by stbergmann** [comment]
bot, build org.libreoffice.LibreOffice.BundledExtension.Voikko

**Comment by stbergmann** [comment]
bot, build org.libreoffice.LibreOffice.BundledExtension.Voikko

**Comment by hfiguiere** [comment]
You need add a `flathub.json` like there: https://github.com/flathub/org.gimp.GIMP.Plugin.GMic/blob/master/flathub.json to tell Flathub to legitimately ignore the lack of icon

**Comment by stbergmann** [comment]
bot, build org.libreoffice.LibreOffice.BundledExtension.Voikko

**Comment by jpakkane** [comment]
Should this work in the released version? There have been several updates to LO since this was merged but when I open a Finnish document in LO it always says "missing hyphenation data". The extension manager does not show the Voikko extension either.,

**Comment by stbergmann** [comment]
The issue that the Voikko extension does not show up on the page of LibreOffice from Flathub in e.g. GNOME Software is being discussed on the Matrix #flatpak channel.  It appears that for now you have to install it manually.

**Comment by hfiguiere** [comment]
extensions are not installed automatically as per the libreoffice manifest :

`            "no-autodownload": true,`

And for GNOME software, it's management of extensions is greatly lacking. I have a few issues open.

**Comment by jpakkane** [comment]
Installing it manually means building from source as upstream does not provide a prebuilt extension for Linux but instead say that "on Linux the extension is provided by the distribution" as explained [on this page](https://voikko.puimula.org/linux-jakelut.html) (though only in Finnish).

**Comment by hfiguiere** [comment]
Now it's visible in GNOME Software:

image

So, problem solved.

**Comment by hfiguiere** [comment]
you still need in install it "manually", ie to opt to install it.

**Comment by hfiguiere** [comment]
Just to clarify, it was fixed by https://github.com/flathub/org.libreoffice.LibreOffice.BundledExtension.Voikko/pull/1

There was something broken.

**Comment by stbergmann** [comment]
@hfiguiere Thanks for spotting and fixing that silly typo!
@jpakkane Just to clarify that with "install it manually" I meant some `flatpak install org.libreoffice.LibreOffice.BundledExtension.Voikko` on the command line; sorry for being too terse there.

**Comment by hfiguiere** [comment]
Or by clicking in GNOME Software as shown above.

**Comment by jpakkane** [comment]
Can confirm that this works now. Thanks.

**Pull request closed without merging**